lighthouse-core/scripts/generate-bundlephobia-database.js

```diff
@@ -0,0 +1,145 @@
+const suggestionsJSON = require('../audits/byte-efficiency/library-suggestions.js').suggestions;
+/** @type string[] */
+/* eslint-disable-next-line max-len */
+const librarySuggestions = Object.keys(suggestionsJSON).map(s => suggestionsJSON[s].concat(s)).flat();
+
+/** @type {Record<string, Record<'lastScraped', number|string> | Record<'repository', string> | Record<string, BundlePhobiaLibrary>>} */
+let database = {};
+if (fs.existsSync(databasePath)) {
+  database = require(databasePath);
+}
+
+/**
+ * Returns true if this library has been scraped from BundlePhobia in the past week.
+ * @param {string} library
+ * @return {boolean}
+ */
+function hasBeenRecentlyScraped(library) {
+  if (!database[library] || database[library].lastScraped === 'Error') return false;
+  return (Date.now() - database[library].lastScraped) / (1000 * 60 * 60 * 24.0) < 7;
+}
+
+/**
+ * Returns true if the object represents valid BundlePhobia JSON.
+ * The version string must not match this false-positive expression: '{number} packages'.
+ * @param {any} library
+ * @return {boolean}
+ */
+function validateLibraryObject(library) {
+  return library.hasOwnProperty('name') &&
+    library.hasOwnProperty('size') &&
+    library.hasOwnProperty('gzip') &&
+    library.hasOwnProperty('description') &&
+    library.hasOwnProperty('repository') &&
+    library.hasOwnProperty('version') &&
+    !library.version.match(/^([0-9]+) packages$/);
+}
+
+/**
+ * Save BundlePhobia stats for a given npm library to the database.
+ * @param {string} library
+ * @param {number} index
+ */
+async function collectLibraryStats(library, index) {
+  return new Promise(async (resolve, reject) => {
+    console.log(`\n◉ (${index}/${librarySuggestions.length}) ${library} `);
+
+    if (hasBeenRecentlyScraped(library)) {
+      console.log(`   ❕ Skipping`);
+      resolve();
+      return;
+    }
+
+    /** @type {Array<BundlePhobiaLibrary>} */
+    const libraries = [];
+    /** @type {string|number} */
+    let lastScraped = Date.now();
+
+    const versions = await getPackageVersionList(library, 10);
+    for (const version of versions) {
+      try {
+        const libraryJSON = await fetchPackageStats(version);
+        if (validateLibraryObject(libraryJSON)) libraries.push(libraryJSON);
+      } catch (e) {
+        console.log(`   ❌ Failed to fetch stats | ${version}`);
+        lastScraped = 'Error';
+      }
+    }
+
+    for (let index = 0; index < libraries.length; index++) {
+      const library = libraries[index];
+
+      database[library.name] = {
+        ...database[library.name],
+        [library.version]: {
+          gzip: library.gzip,
+        },
+      };
+
+      if (index === 0) {
+        database[library.name] = {
+          repository: library.repository,
+          lastScraped,
+          ...database[library.name],
+        };
+
+        database[library.name]['latest'] = database[library.name][library.version];
+      }
+
+      if (lastScraped === 'Error') {
```

Comment:
what circumstances does this error occur? without knowing much it feels like we should not save this and throw instead :) but maybe there is a case where we don't want to be alerted to this situation to remove the dead library?

Reply:
This is for when the network request to BundlePhobia times out due to excessive requests, I manually set lastScraped to `Error`. It happened when I had like 49 libraries in the `library-suggestions.json`, so I just had to wait a few min and re-run it. The reason it's still saved is because some of the versions may not have error'd out so we want to keep those just in case. It could also error out if it's a dead library (meaning no longer on npm?) in which case we should remove it from the database. 

I'll add a unit test that makes sure no library has an `Error` value for lastScraped so we know for sure :)